src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumeManager.java

```diff
@@ -0,0 +1,126 @@
+
+  // Reference to the external volume per table
+  private final Map<String, StreamingIngestStorage> externalVolumeMap;
+  private final SnowflakeStreamingIngestClientInternal<T> owningClient;
+
+  private final boolean isTestMode;
+
+  /**
+   * Constructor for ExternalVolumeManager
+   *
+   * @param isTestMode whether the manager in test mode
+   * @param client the owning client
+   */
+  ExternalVolumeManager(boolean isTestMode, SnowflakeStreamingIngestClientInternal<T> client) {
+    this.owningClient = client;
+    this.isTestMode = isTestMode;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Given a blob, return the target storage by looking up the table name from the channel context
+   *
+   * @param blobData the blob to upload
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage getStorage(List<List<ChannelData<T>>> blobData) {
+    // Only one chunk per blob in Iceberg mode.
+    ChannelFlushContext channelContext = blobData.get(0).get(0).getChannelContext();
+    StreamingIngestStorage stage =
+        this.externalVolumeMap.get(channelContext.getFullyQualifiedTableName());
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
+          String.format(
+              "No storage found for table %s", channelContext.getFullyQualifiedTableName()));
+    }
+
+    return stage;
+  }
+
+  /**
+   * Add a storage to the manager by looking up the table name from the open channel response
+   *
+   * @param openChannelResponse response from open channel
+   */
+  @Override
+  public void addStorage(OpenChannelResponse openChannelResponse) {
+    String fullyQualifiedTableName =
+        String.format(
+            "%s.%s.%s",
+            openChannelResponse.getDBName(),
+            openChannelResponse.getSchemaName(),
+            openChannelResponse.getTableName());
+    if (!this.externalVolumeMap.containsKey(fullyQualifiedTableName)) {
+      try {
+        ConfigureCallHandler configureCallHandler =
+            ConfigureCallHandler.builder(
+                    this.owningClient.getHttpClient(),
+                    this.owningClient.getRequestBuilder(),
+                    STREAMING_CHANNEL_CONFIGURE,
+                    CHANNEL_CONFIGURE_ENDPOINT)
+                .setRole(this.owningClient.getRole())
+                .setDatabase(openChannelResponse.getDBName())
+                .setSchema(openChannelResponse.getSchemaName())
+                .setTable(openChannelResponse.getTableName())
+                .build();
+        this.externalVolumeMap.put(
+            fullyQualifiedTableName,
+            new StreamingIngestStorage(
+                isTestMode,
+                configureCallHandler,
+                this.owningClient.getName(),
+                DEFAULT_MAX_UPLOAD_RETRIES));
+      } catch (SnowflakeSQLException | IOException err) {
+        throw new SFException(err, ErrorCode.UNABLE_TO_CONNECT_TO_STAGE);
```

Comment:
1. what happens when this exception is thrown? Does the customer's code get an opportunity to handle this exception, or does it go unhandled and cause a process crash?
2. Lets add retry handling to all API calls made via the new SnowflakeServiceClient I proposed in another thread. Need retry with backoff and jitter.
3. Can we log to whatever log provider is used by the sdk before throwing?

Reply:
Added log for this and retry logic is in `executeWithRetries`. If the error is thrown, the process stop immediately.